### PR TITLE
Feature/4 questions and answers

### PR DESCRIPTION
**Pull request by lisset-orozco**
- [x] question model

- [x] nested resources

- [x] build questions´s list with factoryBot

- [x] create questions

- [x] tests for the creation of questions

- [x] validate owner of the poll

- [x] update and eliminate questions

- [x] action show for questions

- [x] refactor polls controller and create answer_model

- [x] create answers

- [x] update answers

- [x] delete answers

### CODE REVIEW

File: app/controllers/api/v1/answers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+module Api
+  module V1
+    # answers controller
+    class AnswersController < ApiV1Controller
+      # create update destroy
+
+      # POST /polls/1/answers
+      def create
+        authenticate
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/answers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        create_answer if @current_user
+        # create_answer unless authenticate_owner('user')
+      end
+
+      # PATCH/PUT /polls/1/answers/1
+      def update
+        authenticate
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/answers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        set_answer
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/answers_controller.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        update_answer if @current_user
+      end
+
+      # DELETE /polls/1/answers/1
+      def destroy
+        authenticate
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/answers_controller.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        set_answer
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/answers_controller.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        @answer = Answer.new(answer_params)
+        return render 'api/v1/answers/show' if @answer.save
+
+        error_message('error', :unprocessable_entity, @answer)
+      end
+
+      def update_answer
+        return unless authenticate_owner('user')
+
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: delete line

File: app/controllers/api/v1/answers_controller.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        return render('api/v1/answers/show') if @answer.update(answer_params)
+
+        error_message('error', :unprocessable_entity, @answer)
+      end
+
+      def destroy_answer
+        return unless authenticate_owner('destroy')
+
```
Comment (joalbertg) [inline app/controllers/api/v1/answers_controller.rb]: delete line

File: app/controllers/api/v1/my_polls_controller.rb [hunk 8]
```diff
@@ -4,26 +4,26 @@ module Api
...
+      # before_action :authenticate, only: %i[create update destroy]
+      # before_action :set_poll, only: %i[show update destroy]
 
       def index
         @polls = MyPoll.all
       end
 
-      def show; end
+      def show
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/my_polls_controller.rb [hunk 9]
```diff
@@ -4,26 +4,26 @@ module Api
...
+      end
 
       def create
-        @poll = @current_user.my_polls.new(my_polls_params)
-        # poll = MyPoll.create(my_polls_params)
-        # poll.user = @curren_user
-        return render 'api/v1/my_polls/show' if @poll.save
+        authenticate
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/my_polls_controller.rb [hunk 10]
```diff
@@ -33,26 +33,43 @@ def destroy
...
+      def my_polls_params
+        params.require(:poll).permit(:title, :description, :expires_at)
+      end
+
+      def set_poll
+        @poll = MyPoll.find_by_id(params[:id])
+      end
+
       def authenticate_owner(type)
+        authenticate
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/my_polls_controller.rb [hunk 11]
```diff
@@ -33,26 +33,43 @@ def destroy
...
+        return unless @current_user
+
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/my_polls_controller.rb [hunk 12]
```diff
@@ -33,26 +33,43 @@ def destroy
...
-      def destroy_poll
-        @poll.destroy
-        render json: { message: 'the indicated poll was eliminated' }
+        error_message('error', :unprocessable_entity, @poll)
       end
 
-      def set_poll
-        @poll = MyPoll.find_by_id(params[:id])
+      def update_poll
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/my_polls_controller.rb [hunk 13]
```diff
@@ -33,26 +33,43 @@ def destroy
...
+        return render('api/v1/my_polls/show') if @poll.update(my_polls_params)
+
+        error_message('error', :unprocessable_entity, @poll)
       end
 
-      def my_polls_params
-        params.require(:poll).permit(:title, :description, :expires_at)
+      def destroy_poll
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/my_polls_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/questions_controller.rb [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  module V1
+    # questions controller
+    class QuestionsController < ApiV1Controller
+      # before_action :authenticate, only: %i[create update destroy]
+      # before_action :set_question, only: %i[show update destroy]
+      # before_action :set_poll
+
+      # GET /polls/:poll_id/questions
+      def index
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/questions_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/questions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        @questions = @poll.questions
+      end
+
+      # GET /polls/1/questions/2
+      def show
+        set_question
```
Comment (joalbertg) [inline app/controllers/api/v1/questions_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/questions_controller.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+      end
+
+      # POST /polls/1/questions
+      def create
+        create_question unless authenticate_owner('user')
+      end
+
+      # PATCH/PUT /polls/1/questions/1
+      def update
+        set_question
```
Comment (joalbertg) [inline app/controllers/api/v1/questions_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/questions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        update_question unless authenticate_owner('user')
+      end
+
+      # DELETE /polls/1/questions/1
+      def destroy
+        set_question
```
Comment (joalbertg) [inline app/controllers/api/v1/questions_controller.rb]: refactor all this methods, apply DRY

File: app/controllers/api/v1/questions_controller.rb [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+      end
+
+      def set_question
+        @question = Question.find_by_id(params[:id])
+      end
+
+      def authenticate_owner(type)
+        return unless authenticate
+
+        set_poll
```
Comment (joalbertg) [inline app/controllers/api/v1/questions_controller.rb]: refactor all this methods, apply DRY

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/answers_controller.rb (hunk 1 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 2 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 3 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 4 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 5 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 6 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/answers_controller.rb (hunk 7 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/answers_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 8 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 9 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 10 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 11 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 12 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/my_polls_controller.rb (hunk 13 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/my_polls_controller.rb]: done

File: app/controllers/api/v1/questions_controller.rb (hunk 14 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/questions_controller.rb]: done

File: app/controllers/api/v1/questions_controller.rb (hunk 15 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/questions_controller.rb]: done

File: app/controllers/api/v1/questions_controller.rb (hunk 16 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/questions_controller.rb]: done

File: app/controllers/api/v1/questions_controller.rb (hunk 17 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/questions_controller.rb]: done

File: app/controllers/api/v1/questions_controller.rb (hunk 18 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/questions_controller.rb]: done

**Review by joalbertg — APPROVED** [review]

**Pull request merged**